# 응용프로그래밍개발 2일차

## 복습하기 과제

### 1번 과제

In [2]:
name = '홍길동'
age = 30
height = 171.5
print(f'성: {name[0]}')
print(f'이름: {name[1:]}')
print(f'{name}의 나이: {age}세, 키: {height}cm')

성: 홍
이름: 길동
홍길동의 나이: 30세, 키: 171.5cm


### 2번 과제

In [4]:
name = input('이름을 입력: ')
year = int(input('현재년도 입력: '))
born = int(input('탄생년도 입력: '))

print(f'{name}님의 나이는 {year - born + 1}세(우리나라 기준) 입니다!')

이현우님의 나이는 18세(우리나라 기준) 입니다!


### 3번 과제

In [10]:
from datetime import datetime

yearNum = int(input('년 입력: '))
monthNum = int(input('월 입력: '))
dateNum = int(input('일 입력: '))

# 1번
print(f'{yearNum}-{monthNum}-{dateNum}')

# 2번 (진도 역행)
target = datetime(year = yearNum, month = monthNum, day = dateNum)
print(target.strftime('%Y-%m-%d (%A)'))

'2021-09-11 (Saturday)'

### 4번 과제

In [11]:
vowels = 'aeiou'

char = input('영어 알파벳을 입력하세요: ')
char2 = char.lower()
isVowel = vowels.find(char2) != -1
if isVowel:
  print(f'{char}은 모음')
else:
  print(f'{char}은 자음')

a은 모음


### 5번 과제

In [13]:
messages = ['Thank you', 'Well done', 'Correct']

while True:
  selection = int(input(f'입력 1, 2, 3: '))
  if not 1 <= selection <= 3:
    print('Error Message')
    break
  print(messages[selection - 1])

Well done
---------------------
Correct
---------------------
Thank you
---------------------
Error Message


## 3. 반복문

### playground

In [2]:
for _ in range(5):
  print('와!')

와!
와!
와!
와!
와!


In [4]:
i = 0
while i < 5:
  print(i, end = ', ')
  i += 1

0, 1, 2, 3, 4, 

In [5]:
for i in range(10):
  print(i, end = ' ')

0 1 2 3 4 5 6 7 8 9 

In [7]:
for i in range(1, 11, 2):
  print(i, end = ' ')

1 3 5 7 9 

In [9]:
for i in range(20, 0, -2):
  print(i, end = ' ')

20 18 16 14 12 10 8 6 4 2 

### 실습 1

In [11]:
phoneNum = input('휴대폰 번호 입력: ')
output = ''

for c in phoneNum:
  if not c == '-':
    output += c

print(output)

01012345678


### 실습 2

In [13]:
original = input('휴대폰 번호 입력: ')
print(f'{original[0:3]}-{original[3:7]}-{original[7:11]}')

010-1234-5678


### 실습 3

In [3]:
print('킬로그램(kb)\t파운드(lb)\t온스(oz)')

for kg in range(10, 101, 5):
  lb = kg * 2.204623
  oz = kg * 35.273962
  print('%8d\t%10.1f\t%10.1f' % (kg, lb, oz))

킬로그램(kb)	파운드(lb)	온스(oz)
      10	      22.0	     352.7
      15	      33.1	     529.1
      20	      44.1	     705.5
      25	      55.1	     881.8
      30	      66.1	    1058.2
      35	      77.2	    1234.6
      40	      88.2	    1411.0
      45	      99.2	    1587.3
      50	     110.2	    1763.7
      55	     121.3	    1940.1
      60	     132.3	    2116.4
      65	     143.3	    2292.8
      70	     154.3	    2469.2
      75	     165.3	    2645.5
      80	     176.4	    2821.9
      85	     187.4	    2998.3
      90	     198.4	    3174.7
      95	     209.4	    3351.0
     100	     220.5	    3527.4


### 실습 4

In [4]:
for i in range(1, 1001):
  print(i, end = ' ')
  if i == 10: break

1 2 3 4 5 6 7 8 9 10 

### 즉석 실습 1

In [9]:
number = 0
for i in range(1, 101):
  number += i

print(number)

5050


In [11]:
number = 0
i = 1
while i <= 100: # while로 바꾸기
  number += i
  i += 1

print(number)

5050


In [12]:
toNumber = int(input('1에서 무엇까지 더할까요? '))
# number = 0
# for i in range(1, toNumber + 1):
#   number += i
number = int(toNumber * (toNumber + 1) / 2) # 성능 최적화

print(number)

5050


### 즉석 실습 2

In [15]:
for i in range(5, 0, -1):
  print(i, end = ' ')

5 4 3 2 1 

### 즉석 실습 3

In [22]:
for column in range(0, 11):
  for row in range(0, 11):
    print((column + row) % 10, end = ' ')
  print()


0 1 2 3 4 5 6 7 8 9 0 
1 2 3 4 5 6 7 8 9 0 1 
2 3 4 5 6 7 8 9 0 1 2 
3 4 5 6 7 8 9 0 1 2 3 
4 5 6 7 8 9 0 1 2 3 4 
5 6 7 8 9 0 1 2 3 4 5 
6 7 8 9 0 1 2 3 4 5 6 
7 8 9 0 1 2 3 4 5 6 7 
8 9 0 1 2 3 4 5 6 7 8 
9 0 1 2 3 4 5 6 7 8 9 
0 1 2 3 4 5 6 7 8 9 0 


## 문제 만들기
어떤 전화번호를 string의 형식('abc' 같은)으로 받았을 때, 그 전화번호가 실제로 타당한 전화번호인지 검증하시오.

검증 조건:
- 가능한 형식에는 `010-1234-5678`(휴대전화), `053-123-4567`(지역 코드 + 번호), `123-4567`(지역 코드가 생략된 번호)이 있다. 여기서 `010`(휴대 전화), `053`(지역 코드) 부분은 반드시 일치하는지 확인해야 한다.  
  특히 가능한 지역 코드에는 두 자리(`02`), 세 자리(`'031', '032', '033', '041', '042', '043', '044', '051', '052', '053', '054', '055', '061', '062', '063', '064'`) 가 있다.
- 받은 텍스트에는 `-`가 포함될 수도, 포함되지 않을 수도 있다. 즉, `01012345678` 처럼 될 수도 있다.

### 예시 정답

In [2]:
localeCodes = ['02', '031', '032', '033', '041', '042', '043', '044', '051', '052', '053', '054', '055', '061', '062', '063', '064']
num = input('전화번호를 입력해주세요: ')

# 1. 편의를 위해 일관적으로 -를 텍스트에서 제거
num2 = ''
for c in num:
  if not c == '-': num2 += c

# 2. 글자 길이에 따라 분류
l = len(num2)
if l == 11: # 휴대전화
  if num2[0:3] == '010':
    print('알맞은 휴대전화 번호입니다.')
  else:
    print('잘못된 휴대전화 번호입니다. 휴대전화 번호는 010으로 시작해야 합니다.')

  
elif l == 10 or l == 9: # 지역 코드(3자리/2자리) + 번호

  # 지역코드 추출
  if l == 10: # 지역 코드(3자리) + 번호
    code = num2[0:3]
  else: # 지역 코드(2자리) + 번호
    code = num2[0:2]

  # 바른 지역코드인지 확인.
  # 코드를 더 길게 쓰기 귀찮아서 이렇게 했지, if code == '02' or code == '031', ... 같은 식으로 해도 괜찮습니다.
  # 사실 지금까지 배운 범위에서는 그게 정석입니다.
  if code in localeCodes:
    print(f'알맞은 지역 전화번호(지역코드: {code})입니다.')
  else:
    print(f'알 수 없는 지역코드({code})입니다.')


elif l == 7: # 지역 코드가 생략된 번호
  print('알맞은 지역 전화번호입니다.')


else: # 바르지 않은 전화번호
  print('알맞지 않은 전화번호입니다.')


알맞은 지역 전화번호입니다.


### 예시 정답 2
- 이 버전은 조건이랑 다르게 전화번호를 `010-1234-5678` 같은 `-`가 붙은 형식으로 바꿔서 출력해줍니다.
- 조금 더 실생활과 가까운 버전입니다.

In [9]:
localeCodes = ['02', '031', '032', '033', '041', '042', '043', '044', '051', '052', '053', '054', '055', '061', '062', '063', '064']
num = input('전화번호를 입력해주세요: ')

# 1. 편의를 위해 일관적으로 -를 텍스트에서 제거
num2 = ''
for c in num:
  if not c == '-':
    num2 += c

# 2. 글자 길이에 따라 분류
l = len(num2)
if l == 11:  # 휴대전화
  if num2[0:3] == '010':
    print(f'휴대전화 번호: {num2[0:3]}-{num2[3:7]}-{num2[7:11]}')
  else:
    print('잘못된 휴대전화 번호입니다. 휴대전화 번호는 010으로 시작해야 합니다.')


elif l == 7:  # 지역 코드가 생략된 번호
  print(f'지역 전화번호: {num2[0:3]}-{num2[3:7]}')


elif num2[0:3] == '010': # 실수로 전화번호를 입력하다 덜 입력했을 수도 있음
  print('휴대전화 번호를 덜 입력하셨습니다. 휴대전화 번호는 010-XXXX-XXXX의 형식으로 입력해야 합니다.')


elif l == 10 or l == 9:  # 지역 코드(3자리/2자리) + 번호


  # 지역코드 추출
  if l == 10:  # 지역 코드(3자리) + 번호
    code = num2[0:3]
    content = num2[3:10]
  else:  # 지역 코드(2자리) + 번호
    code = num2[0:2]
    content = num2[3:9]

  # 바른 지역코드인지 확인.
  # 코드를 더 길게 쓰기 귀찮아서 이렇게 했지, if code == '02' or code == '031', ... 같은 식으로 해도 괜찮습니다.
  # 사실 지금까지 배운 범위에서는 그게 정석입니다.
  if code in localeCodes:
    print(f'지역 전화번호: ({code}) {content[0:3]}-{content[3:7]}')
  else:
    print(f'알 수 없는 지역코드({code})입니다.')


else:  # 바르지 않은 전화번호
  print('알맞지 않은 전화번호입니다.')


잘못된 휴대전화 번호입니다. 휴대전화 번호는 010으로 시작해야 합니다.
